contracts/src/contracts/ValidatorManager/BalancerValidatorManager.sol

```diff
@@ -0,0 +1,320 @@
+        }
+        if ($.validatorPendingWeightUpdate[validationID] == 0) {
+            revert BalancerValidatorManager__NoPendingWeightUpdate(validationID);
+        }
+
+        if (validator.messageNonce == 0) {
+            revert InvalidValidationID(validationID);
+        }
+
+        // Submit the message to the Warp precompile.
+        WARP_MESSENGER.sendWarpMessage(
+            ValidatorMessages.packL1ValidatorWeightMessage(
+                validationID, validator.messageNonce, validator.weight
+            )
+        );
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function getChurnPeriodSeconds() external view returns (uint64 churnPeriodSeconds) {
+        return _getChurnPeriodSeconds();
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function getSecurityModules() external view returns (address[] memory securityModules) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        return $.securityModules.keys();
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function getSecurityModuleWeights(
+        address securityModule
+    ) external view returns (uint64 weight, uint64 maxWeight) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        weight = $.securityModuleWeight[securityModule];
+        maxWeight = uint64($.securityModules.get(securityModule));
+
+        return (weight, maxWeight);
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function isValidatorPendingWeightUpdate(
+        bytes32 validationID
+    ) external view returns (bool) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        return $.validatorPendingWeightUpdate[validationID] != 0;
+    }
+
+    function _checkValidatorSecurityModule(
+        bytes32 validationID,
+        address securityModule
+    ) internal view {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        // If the validator has no associated security module, it is an initial validator
+        // and can be managed by any security module
+        if ($.validatorSecurityModule[validationID] == address(0)) {
+            return;
+        } else if ($.validatorSecurityModule[validationID] != securityModule) {
+            revert BalancerValidatorManager__ValidatorNotBelongingToSecurityModule(
+                validationID, securityModule
+            );
+        }
+    }
+
+    function _setupSecurityModule(address securityModule, uint64 maxWeight) internal {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        if (maxWeight == 0) {
+            if (!$.securityModules.remove(securityModule)) {
+                revert BalancerValidatorManager__SecurityModuleNotRegistered(securityModule);
+            }
+        } else {
+            $.securityModules.set(securityModule, uint256(maxWeight));
```

Comment:
Shouldn't we increment `$.l1TotalWeight` by `maxWeight` here?

Reply:
No. `l1TotalWeight` is supposed to be the current total weight of all validators.